File: src/main/java/custom/tibame201020/adbOpenCv/opencv/ocr/CharacterOCR.java
package custom.tibame201020.adbOpenCv.opencv.ocr;

import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.Rect;
import org.opencv.imgproc.Imgproc;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

@Service
public class CharacterOCR implements MatchOCRAdaptor {

    record MatchResult(
            String character,
            double x
    ) {
    }

    @Override
    public String executeOCR(String ocrTemplatesPath, Mat targetImg, double threshold) throws Exception {
        var templates = loadTemplateMatMap(ocrTemplatesPath);

        List<MatchResult> results = new ArrayList<>();

        for (Map.Entry<String, Mat> entry : templates.entrySet()) {
            String character = entry.getKey();
            Mat template = entry.getValue();

            Mat res = new Mat();
            Imgproc.matchTemplate(targetImg, template, res, Imgproc.TM_CCOEFF_NORMED);

            // 將 res 中 >= threshold 的位置設為 255，其餘設為 0
            Mat mask = new Mat();
            Imgproc.threshold(res, mask, threshold, 255, Imgproc.THRESH_BINARY);

            // 轉成 8-bit 單通道
            mask.convertTo(mask, CvType.CV_8U);

            // 找出所有匹配區塊
            List<MatOfPoint> contours = new ArrayList<>();
            Mat hierarchy = new Mat();
            Imgproc.findContours(mask, contours, hierarchy, Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE);

            for (MatOfPoint contour : contours) {
                Rect rect = Imgproc.boundingRect(contour);
                results.add(new MatchResult(character, rect.x));
            }
        }

        // 根據 x 座標排序
        results.sort(Comparator.comparingDouble(r -> r.x));

        StringBuilder sb = new StringBuilder();
        for (MatchResult res : results) {
            sb.append(res.character());
        }
        return sb.toString();
    }
}

File: core/service/core/opencv/ocr/character_ocr.py
"""
字元 OCR - 用於辨識字元
"""
import cv2
import numpy as np
from typing import List, Tuple
from .match_ocr_adaptor import MatchOCRAdaptor


class CharacterOCR(MatchOCRAdaptor):
    """字元 OCR 類別"""
    
    def execute_ocr(self, ocr_templates_path: str, target_img: np.ndarray, threshold: float) -> str:
        """
        執行字元 OCR
        
        Args:
            ocr_templates_path: OCR 模板基礎路徑
            target_img: 目標影像
            threshold: 閾值
            
        Returns:
            辨識出的字串
        """
        templates = self.load_template_mat_map(ocr_templates_path)
        
        if not templates:
            return ""
        
        results: List[Tuple[str, float]] = []  # (character, x)
        
        for character, template in templates.items():
            # 確保模板尺寸不超過目標影像
            if template.shape[0] > target_img.shape[0] or template.shape[1] > target_img.shape[1]:
                continue
            
            res = cv2.matchTemplate(target_img, template, cv2.TM_CCOEFF_NORMED)
            
            # 將 >= threshold 的位置設為 255，其餘設為 0
            _, mask = cv2.threshold(res, threshold, 255, cv2.THRESH_BINARY)
            mask = mask.astype(np.uint8)
            
            # 找出所有匹配區塊
            contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
            
            for contour in contours:
                x, y, w, h = cv2.boundingRect(contour)
                results.append((character, float(x)))
        
        # 根據 x 座標排序
        results.sort(key=lambda r: r[1])
        
        # 組合字串
        return ''.join([r[0] for r in results])

